lib/msprime.c

```diff
@@ -4698,6 +4706,104 @@ dtwf_model_rate_to_generation_rate(simulation_model_t *MSP_UNUSED(model), double
     return rate;
 }
 
+/**************************************************************
+ * Allele frequency trajectory simulation for genic selection
+ *
+ **************************************************************/
+
+static double
+genic_selection_stochastic_forwards(double dt, double freq, double alpha, double u)
+{
+    double ux = (alpha * freq * (1 - freq)) / tanh(alpha * freq);
+    int sign = u < 0.5? 1: -1;
+    return freq + (ux * dt) + sign * sqrt(freq * (1.0 - freq) * dt);
+}
+
+static int
+genic_selection_generate_trajectory(sweep_t *self, msp_t *simulator,
+        size_t *ret_num_steps, double **ret_time, double **ret_allele_frequency)
+{
+    int ret = 0;
+    genic_selection_trajectory_t trajectory =
+        self->trajectory_params.genic_selection_trajectory;
+    gsl_rng *rng = simulator->rng;
+    size_t max_steps = 64;
+    double *time = malloc(max_steps * sizeof(*time));
+    double *allele_frequency = malloc(max_steps * sizeof(*allele_frequency));
+    double x, t, *tmp;
+    size_t num_steps;
+    double current_size = 1.0;
+
+    if (time == NULL || allele_frequency == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    /* TODO Wrap this in a rejection sample loop and get the population size
+     * from the simulator. We can use
+     * pop_size = get_population_size(sim->populations[0], time);
+     * to do this because we assume there are no demographic events
+     * during a sweep */
+
+    x = trajectory.end_frequency;
+    t = simulator->time;
+    num_steps = 0;
+
+    while (x > trajectory.start_frequency) {
+        if (num_steps + 1 >= max_steps) {
+            max_steps *= 2;
+            tmp = realloc(time, max_steps * sizeof(*time));
+            if (tmp == NULL) {
+                ret = MSP_ERR_NO_MEMORY;
+                goto out;
+            }
+            time = tmp;
+            tmp = realloc(allele_frequency, max_steps * sizeof(*allele_frequency));
+            if (tmp == NULL) {
+                ret = MSP_ERR_NO_MEMORY;
+                goto out;
+            }
+            allele_frequency = tmp;
+        }
+        time[num_steps] = t;
+        allele_frequency[num_steps] = x;
+        x = 1.0 - genic_selection_stochastic_forwards(
+                trajectory.dt, 1.0 - x, trajectory.alpha * current_size,
+                gsl_rng_uniform(rng));
+        t += trajectory.dt;
+        num_steps++;
+    }
+    assert(num_steps < max_steps); /* num_steps + 1 above guarantees this */
+    /* JK: question to Andy: Does this bookending make sense?
```

Comment:
yeah i think this makes sense. need to have a harder think about it, but my knee jerk impression is this should work fine